activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -169,6 +169,10 @@ def connection_retries
         (@config[:connection_retries] || 1).to_i
       end
 
+      def connection_retry_timeout
+        (@config[:connection_retry_timeout] || 5).to_f
```

Comment:
I wonder if instead of defaulting to an arbitrary 5 seconds, we should derive the default from the read timeout ? 

e.g. `connection_retry_timeout || read_timeout * 0.95` or something.

Reply:
That sounds like a saner default! 👍 I'll make the change 😄

Reply:
Hm, actually I might be missing something @byroot -- where does the read timeout come from? I mistakenly thought the `timeout` from the db config was a read timeout, but I've just realized that's the timeout to check out a connection from the pool 😅

Reply:
Right, I'm not sure if it's normalized for all adapters, I'll check.

Reply:
I think I kinda prefer an arbitrary 5 seconds... I'd rather not conflate it with network-behaviour expectations, where [especially as a global default] this is more of a failsafe "if we've already spent a while on it, best not retry and risk doubling that time before we return a failure".

Reply:
Hum, so Active Record isn't actually aware of the timeout :/

For MySQL it's passed transparently to the client, as `read_timeout`:
  - https://github.com/rails/rails/blob/5fb3e721e70a13a7c0facf2d9175c15fb5132cc4/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb#L30
  - https://github.com/brianmario/mysql2#connection-options

For Postgres I actually couldn't find any reference to a timeout.

So I guess we can stay with an arbitrary default like @matthewd suggest, but IMO 5 seconds is HUGE. Most connection errors will either happen very fast or will hang forever. If it was just me I'd suggest something like `100ms`, but I suppose something more conservative like `1s` might be preferable.

Reply:
Yeah, fair! If 1s is okay with Matthew, I'll set that to the default.

Reply:
We're not defining a timeout here, though: we're judging how long elapsed before something else reported a failure, and using that to decide whether to give up without retrying _an otherwise seemingly retryable_ operation. If something took a couple of seconds to get through a few stack layers and then ultimately itself timeout [having decided its other end has gone away] and report a retryable error, the whole point of this functionality is that we then want to retry it.

I see this as a last-ditch attempt to avoid wasting so much time that we timeout the web request... if your DB driver is reporting a retryable error and you can't afford to risk waiting its timeout a second time, that's the timeout you should be lowering.

One second is not a reasonable timeframe to handle even trivial misbehaviour on a remote DB connection: by the time any lower layer has decided something went wrong and recommended that we reconnect and retry, we'll have crossed our threshold and disabled retry behaviour.

I'm on the fence about whether we should have a non-nil value for this at all, tbh... I'd want something to be absolutely catastrophically wrong for us to bypass the whole retry behaviour that we've just added. Even 5s feels aggressive for that (notably, that's the value that's been in use at Shopify, where I imagine there is far more specific understanding of production network behaviours than in the average app), but it strikes me as a reasonable middle ground -- mostly because, as you note, nothing's ever actually going to be taking +/- 5 seconds: either it'll be much faster, or the lower-layer timeout will be much longer and the elapsed time we see will be in the 10+ or 60+ sec range, which is indeed "that's too broken to even consider retrying" territory.

Reply:
> One second is not a reasonable timeframe to handle even trivial misbehaviour on a remote DB connection: by the time any lower layer has decided something went wrong and recommended that we reconnect and retry, we'll have crossed our threshold and disabled retry behaviour.

That's not really the point though. 

One of the core resiliency principle to avoid cascading failures, if to have fast failures, so it's a bad idea to retry a slow failure.

`1s` is over a hundred time more than what is necessary to get a "connection reset by peer" and similar common event when the connection became invalid while it was sitting in the pool, because the server closed the connection or similar.

Also keep in mind that for most web apps out there, the response time budget is in the 100ms range, so if you were stuck on a DB operation that took over 1s you might as well give up now to preserve the platform.

> I'm on the fence about whether we should have a non-nil value for this at all, tbh...

I'm fine with making this a optional feature if we can't agree on a default. But I think it would be good to have a reasonable default for the vast majority of users who won't know how to tune this (or that it even exists).

Reply:
> perhaps we should avoid calling this a "timeout",

Yeah, you can call it `retry_deadline`.

Reply:
> for most web apps out there, the response time budget is in the 100ms range

Hrmmm, I believe most web apps have request timeout configured at 60 seconds -- 30 if they're being conservative. 

The most _popular_ apps, certainly not... but I'm far more concerned about our default config working acceptably for a near-zero-traffic app that's using a DB in a different cloud provider from their web app than I am about ensuring it's tuned to fail fast for an app with a substantial dedicated ops team. For a start, the former is far more likely to encounter recoverable connection-related errors.

If 5s has been acceptable at Shopify's scale for some time, I'm really struggling to buy that Average Joe's Web App needs a substantially tighter limit for resiliency.

Reply:
I guess to be more specific with an example: [in the future, after we teach more queries that they're definitionally idempotent and thus safe to retry] if we encounter an issue a couple of seconds into the DB server running a pretty-slow SQL query, I would want this code to retry. If that's too long of a query runtime for a given app, I would expect them to have configured a query timeout at a more appropriate level (which will in turn result in a non-retryable error).

Reply:
> I believe most web apps have request timeout configured at 60 seconds -- 30 if they're being conservative.

I'm not talking about overall request timeout, but response time budget, as in the response time you are trying to reach for the vast majority of your requests (e.g. p90).

As for request timeout, most apps use Puma, so they don't have an overall request timeout at all.

> If 5s has been acceptable at Shopify's scale for some time

It doesn't matter that much for Shopify because we have a strict semian config, so slow failure are quickly turned into fast failures. The reason for the 5s is that our monkey patch when it was introduced was quite dumb and wouldn't differentiate between query timeout and other types of errors. That 5s was basically a way to single out timeouts.

> Average Joe's Web App needs a substantially tighter limit for resiliency.

What about we name it `retry_deadline`, default it to `nil` (disabled) for now, as to unblock Adrianna, and then if any of us feel strongly about having it enabled by default we debate this in a followup PR?

Reply:
> It doesn't matter that much for Shopify because we have a strict semian config, so slow failure are quickly turned into fast failures.

That's fair. (I do have my eye on teaching AR some sort of circuit-breaking behaviour as a counterpoint to these retries at some point, incidentally, but that's not particularly relevant here atm.)

> What about we name it retry_deadline, default it to nil (disabled) for now, as to unblock Adrianna, and then if any of us feel strongly about having it enabled by default we debate this in a followup PR?

Yep, that works for me; much of the reason that I was okay with 5s was the notion that it would be long enough that it would be functionally disabled in all but the most egregious of situations.